### PR TITLE
PUD-98: Add caseworker band to User Details. Filter recalls based on band

### PR DESCRIPTION
**Pull request by philsegal**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt [hunk 1]
```diff
@@ -103,13 +105,76 @@ class RecallControllerTest {
...
+    val results = underTest.findAll(bearerToken)
+
+    assertThat(results.size, equalTo(7))
 
-    val results = underTest.findAll()
+    assertThat(
+      results,
+      List<RecallResponse>::containsAll,
+      listOf(
+        RecallResponse(beingBookedOnRecall.recallId(), beingBookedOnRecall.nomsNumber, createdByUserId, now, now, FirstName("Barrie"), null, LastName("Badger"), NameFormatCategory.FIRST_LAST, Status.BEING_BOOKED_ON),
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt:139]: Building `expected` like this makes it pretty hard to read - "to see the wood for the trees".
Can we e.g. move the duplication to one or more helper methods in this test please?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt [hunk 2]
```diff
@@ -103,13 +105,76 @@ class RecallControllerTest {
...
+
+    val results = underTest.findAll(bearerToken)
+
+    assertThat(results.size, equalTo(5))
+
+    assertThat(
+      results,
+      List<RecallResponse>::containsAll,
+      listOf(
+        RecallResponse(beingBookedOnRecall.recallId(), beingBookedOnRecall.nomsNumber, createdByUserId, now, now, FirstName("Barrie"), null, LastName("Badger"), NameFormatCategory.FIRST_LAST, Status.BEING_BOOKED_ON),
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt:170]: Ditto of course.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt [hunk 3]
```diff
@@ -103,13 +105,76 @@ class RecallControllerTest {
     assertThat(results.body, equalTo(recallResponse.copy(recallId = recall.recallId(), createdByUserId = userUuid)))
   }
 
+  private val bookedByUserId = ::UserId.random()
+  private val assignee = ::UserId.random()
+  private val dossierCreatedByUserId = ::UserId.random()
+
+  private val beingBookedOnRecall = Recall(::RecallId.random(), NomsNumber("lf34546"), createdByUserId, now, FirstName("Barrie"), null, LastName("Badger"))
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt:112]: Pretty hard to see wood for trees here.
Created these Recall instances in a test helper method?
Also do the NOMS numbers need any relation to one another?  If not then they can all be randomNoms?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt [hunk 4]
```diff
@@ -103,13 +105,76 @@ class RecallControllerTest {
...
+        RecallResponse(dossierInProgressRecall.recallId(), dossierInProgressRecall.nomsNumber, createdByUserId, now, now, FirstName("Barrie"), MiddleNames("Barnie"), LastName("Badger"), NameFormatCategory.FIRST_MIDDLE_LAST, Status.DOSSIER_IN_PROGRESS, recallNotificationEmailSentDateTime = now, assignee = assignee, assigneeUserName = "Mickey Mouse"),
+        RecallResponse(dossierIssuedRecall.recallId(), dossierIssuedRecall.nomsNumber, createdByUserId, now, now, FirstName("Barrie"), MiddleNames("Barnie"), LastName("Badger"), NameFormatCategory.FIRST_MIDDLE_LAST, Status.DOSSIER_ISSUED, dossierCreatedByUserId = dossierCreatedByUserId)
+      )
+
+    )
+  }
 
-    assertThat(results, equalTo(listOf(recallResponse)))
+  @Test
+  fun `gets all recalls for a band THREE returns only appropriate recalls`() {
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt:152]: Should the test name identify what the 'appropriate' recalls are?  That would help to document the feature.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/AuthenticatedClient.kt [hunk 5]
```diff
@@ -35,7 +35,7 @@ class AuthenticatedClient(
   private val jwtAuthenticationHelper: JwtAuthenticationHelper
 ) {
 
-  var userId: UserId = ::UserId.random()
+  val userId: UserId = ::UserId.random()
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/AuthenticatedClient.kt:38]: 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/UserController.kt [hunk 6]
```diff
@@ -57,18 +58,36 @@ class UserController(
...
+  )
 
 data class AddUserDetailsRequest(
   val firstName: FirstName,
   val lastName: LastName,
   val signature: String,
   val email: Email,
-  val phoneNumber: PhoneNumber
+  val phoneNumber: PhoneNumber,
+  val caseworkerBand: CaseworkerBand = CaseworkerBand.FOUR_PLUS // TODO remove defaulting once UI change released
```
Comment (AndyNewton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/UserController.kt:90]: Would be nice to have a ticket Id in the // TODO

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallController.kt [hunk 7]
```diff
@@ -334,14 +342,18 @@ data class Pdf(val content: String) {
...
-  IN_ASSESSMENT,
-  RECALL_NOTIFICATION_ISSUED,
-  DOSSIER_IN_PROGRESS,
-  DOSSIER_ISSUED,
-  STOPPED
+
+private val ALL_BANDINGS = setOf(THREE, FOUR_PLUS)
+private val FOUR_PLUS_ONLY = setOf(FOUR_PLUS)
+
+enum class Status(val bandings: Set<CaseworkerBand>) {
```
Comment (AndyNewton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallController.kt:349]: I find the term `bandings`, err, a bit 🤯 .  It doesn't tell me anything of what it means TBH.  Even just `visibility` would be preferable?
Perhaps `visibilityBands`?

**Review by AndyNewton — DISMISSED** [review]
OK as is if you like but please take a look at comments either before or after.  Ta.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/UserController.kt (hunk 6 above)
Comment (philsegal) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/UserController.kt:90]: Not moving the back end ticket to done until this is done.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt [hunk 8]
```diff
@@ -103,13 +101,77 @@ class RecallControllerTest {
...
+      )
+    )
+  }
+
   @Test
-  fun `gets all recalls`() {
-    every { recallRepository.findAll() } returns listOf(recall)
+  fun `gets all recalls for a band THREE returns only recalls in BEING_BOOKED_ON, STOPPED, RECALL_NOTIFICATION_ISSUED, DOSSIER_IN_PROGRESS, DOSSIER_ISSUED statuses`() {
+    val bearerToken = "Bearer header.payload"
+    every { recallRepository.findAll() } returns listOf(beingBookedOnRecall, bookedOnRecall, inAssessmentRecall, stoppedRecall, recallNotificationIssuedRecall, dossierInProgressRecall, dossierIssuedRecall)
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/RecallControllerTest.kt:151]: No need now but I think I'd prob create a val list `allRecallStates` for this `listOf` and re-use across the two tests.  Also prob the other unchanged stubbing too.  All just to emphasise that the diff is just from the CW band.

**Review by AndyNewton — APPROVED** [review]
Left one more comment on the tests but can e.g. follow later.

**Pull request merged**